File: src/print_partner/core/path_tree.py
"""Build nested directory trees from relative file paths."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class PathTreeNode:
    """Directory node: ``path`` is the full relative directory ('' for repo root)."""

    name: str
    path: str
    subdirs: dict[str, PathTreeNode] = field(default_factory=dict)
    files: list[str] = field(default_factory=list)


def iter_path_segments(relative_path: str) -> tuple[list[str], str]:
    """Return directory segments and filename for a posix-style relative path."""
    normalized = relative_path.replace("\\", "/").strip("/")
    if not normalized:
        return [], ""
    parts = normalized.split("/")
    if len(parts) == 1:
        return [], parts[0]
    return parts[:-1], parts[-1]


def ensure_subdir(parent: PathTreeNode, segment: str) -> PathTreeNode:
    path = f"{parent.path}/{segment}" if parent.path else segment
    if segment not in parent.subdirs:
        parent.subdirs[segment] = PathTreeNode(name=segment, path=path)
    return parent.subdirs[segment]


def build_path_tree(relative_paths: list[str]) -> PathTreeNode:
    """Nest relative paths into a directory tree (files stored on their parent folder node)."""
    root = PathTreeNode(name="", path="")
    for rel in relative_paths:
        dir_parts, _filename = iter_path_segments(rel)
        parent = root
        for seg in dir_parts:
            parent = ensure_subdir(parent, seg)
        parent.files.append(rel)
    return root

File: web/packages/domain/src/path-tree.ts
/** Build nested directory trees from relative file paths (ported from Python path_tree.py). */

export type PathTreeNode = {
  name: string;
  path: string;
  subdirs: Map<string, PathTreeNode>;
  files: string[];
};

export function iterPathSegments(relativePath: string): [string[], string] {
  const normalized = relativePath.replace(/\\/g, "/").replace(/^\/+|\/+$/g, "");
  if (!normalized) return [[], ""];
  const parts = normalized.split("/");
  if (parts.length === 1) return [[], parts[0]];
  return [parts.slice(0, -1), parts.at(-1) ?? ""];
}

function ensureSubdir(parent: PathTreeNode, segment: string): PathTreeNode {
  const path = parent.path ? `${parent.path}/${segment}` : segment;
  let sub = parent.subdirs.get(segment);
  if (!sub) {
    sub = { name: segment, path, subdirs: new Map(), files: [] };
    parent.subdirs.set(segment, sub);
  }
  return sub;
}

export function buildPathTree(relativePaths: string[]): PathTreeNode {
  const root: PathTreeNode = { name: "", path: "", subdirs: new Map(), files: [] };
  for (const rel of relativePaths) {
    const [dirParts] = iterPathSegments(rel);
    let parent = root;
    for (const seg of dirParts) {
      parent = ensureSubdir(parent, seg);
    }
    parent.files.push(rel);
  }
  return root;
}
